## Business Analytics BUSA8001- Applied Predictive Analytics
## S2 2025
##  *Programming Task 1* {-}


---

**Assignment Points**: 100  (*worth 30% of total Grade*)

**Due Date**: Friday Week 8 (19 September 2025) @ 11.59pm  
**Submission**: Provide your answers in this Jupypter notebook and submit it via iLearn link  



## **Predictive Analysis of Inventory Dead stock**

### Objective: 
- This assignment focuses on a dataset of Inventory dead stock.
- The primary goal is to predict which stock items are likely to be dead (e.g. not used or moved in a long time) using various data mining methods.

### Background: 
A mid manufacturing retailer noticed that certain items in its warehouse across Australia had not moved for a number of months. These products included a number of products or materials stored or manufactured in the wharehouses. Despite being listed in the inventory system, they had no recent sales activity, were no longer featured in marketing campaigns, and had limited relevance to current customer needs.

The inventory manager initiated a review to determine whether these items should be classified as dead stock—inventory that is unlikely to be sold due to obsolescence, lack of demand, damage, or other reasons that can impact their classification. The classification would allow the company to make informed decisions about markdowns, liquidation, or disposal, and free up valuable warehouse space. This project aims to refine this classification by identifying stock items (and possibly their wharehouses locations) that are likely to be dead stock, enhancing the precision of wharhouse facility storage planning

Target variable
- Dead stock (1=yes, 0=no)

The dataset contains the features listed in the second **"Readme"** sheet of the Dataset file





---

**Problem 1** - Reading the dataset (Total Marks: 20)



**Q1**. Create a pandas dataframe contining the first 3,000 rows from the Inventory dataset provided in the **Dataset** folder 
- Delete 'Item No.' column
- Print `info()` of the dataframe

(2 marks)

- Do you recommend dropping/deleting any another column(s) from the Inventory dataset? 

If **YES** then:
1- Identify which one(s) and
2- Justify your selection in one sentence

If **NO** then Justify your answer

(3 marks) 




In [3]:
# ---- provide your code here -----
# Loading pandas and giving it the alias pd
import pandas as pd

# Loading the dataset and keeping only the first 3,000 rows
df = pd.read_excel("~/Desktop/Programming_Task1_S2_2025/Dataset/Inventory.xlsx").iloc[:3000].copy()

# Dropping the identifier column and item description 
df.drop(columns=["Item No. "], inplace=True)
df.drop(columns=["Item Description"], inplace=True)

# Printing structure of the dataframe
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 31 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Whse                              3000 non-null   object 
 1   State                             3000 non-null   object 
 2   Base Unit                         3000 non-null   object 
 3   Total - Quantity                  3000 non-null   float64
 4   Inventory Aging Report Unit Cost  3000 non-null   float64
 5   Total - Value                     3000 non-null   float64
 6   6 Months QTY                      3000 non-null   float64
 7   12 Months QTY                     3000 non-null   float64
 8   2 Years QTY                       3000 non-null   float64
 9   Over 2 Years Qty                  3000 non-null   float64
 10  Over 3 Years Quantity             3000 non-null   float64
 11  Business Area                     3000 non-null   object 
 12  Item T

---- provide your text answer here ----
Yes, I recommend also dropping 'Item Description' because it is a long free-text field that does not provide predictive value for modeling and only adds noise to the dataset.

<hr style="width:25%;margin-left:0;"> 

**Q2**. List which **features** are *numeric*, *ordinal*, and *nominal* variables, and how many features of each kind there are in the dataset.
To answer this question 

- Find the definitions of numeric, ordinal and nominal variables in the course material    
- Carefully consider what values each feature can take as well as the output of `df.info()`.
- Make sure to check the **Readme** sheet in the dataset to understand the features description  

Your answer should be written up in Markdown and include:
1) A table listing all the features present in the dataset and their type (fill out the table template provided below) and
2) A brief description of the contents of the table.

|Variable Kind|Number of Features|Feature Names
| --- | --- | --- |
| Numeric | some number | some text |
| some text  | some number | some text |
| some text  | some number | some text |


(10 marks)

In [2]:
# ---- provide your code here -----
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 31 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Whse                              3000 non-null   object 
 1   State                             3000 non-null   object 
 2   Base Unit                         3000 non-null   object 
 3   Total - Quantity                  3000 non-null   float64
 4   Inventory Aging Report Unit Cost  3000 non-null   float64
 5   Total - Value                     3000 non-null   float64
 6   6 Months QTY                      3000 non-null   float64
 7   12 Months QTY                     3000 non-null   float64
 8   2 Years QTY                       3000 non-null   float64
 9   Over 2 Years Qty                  3000 non-null   float64
 10  Over 3 Years Quantity             3000 non-null   float64
 11  Business Area                     3000 non-null   object 
 12  Item T

---- provide your text answer here ----
| Variable Kind | Number of Features | Feature Names |
|---------------|--------------------|---------------|
| Numeric       | 24                 | Total - Quantity, Inventory Aging Report Unit Cost, Total - Value, 6 Months QTY, 12 Months QTY, 2 Years QTY, Over 2 Years Qty, Over 3 Years Quantity, Jan, Feb, Mar, Apr, May, Jun, Jul, Aug, Sep, Oct, Nov, Dec, Avg monthly, Inventory Turn, Over 2 years Qty, % of over 2 year |
| Nominal       | 6                  | ABC Class, Whse, State, Base Unit, Business Area, Item Type |

The table shows that most features (24) are numeric, measuring stock quantities, costs, values, and turnover. The remaining features (6) are nominal, representing categorical labels such as warehouse, state, business area, item type, and ABC Class. No features are truly ordinal, as ABC Class mixes categories that are not strictly ranked.

<hr style="width:25%;margin-left:0;"> 

**Q3.** Missing Values. 

- Print out the number of missing values for each variable in the dataset and comment on your findings.

(5 marks)

In [5]:
# ---- provide your code here -----
# Checking number of missing values for each column
print(df.isnull().sum())

Whse                                0
State                               0
Base Unit                           0
Total - Quantity                    0
Inventory Aging Report Unit Cost    0
Total - Value                       0
6 Months QTY                        0
12 Months QTY                       0
2 Years QTY                         0
Over 2 Years Qty                    0
Over 3 Years Quantity               0
Business Area                       0
Item Type                           0
ABC Class                           0
Jan                                 0
Feb                                 0
Mar                                 0
Apr                                 0
May                                 0
Jun                                 0
Jul                                 0
Aug                                 0
Sep                                 0
Oct                                 0
Nov                                 0
Dec                                 0
Avg monthly 

---- provide your text answer here ----
The results show that no variables in the dataset contain missing values.

---
---

**Problem 2.** Cleaning data and dealing with categorical features (Total Marks: 40)


**Q1.** 

- Use an appropriate `pandas` function to impute missing values using one of the following two strategies: `mean` and `mode`. (10 marks)
    - Take into consideration the type of each variable (as in Q2 above) and the best practices we discussed in class/lecture notes
- Explain what data imputation is, how you have done it here, and what decisions you had to make. (5 marks)


(Total: 15 marks)

In [6]:
# ---- provide your code here -----
# Filling missing values in numeric columns with mean
for col in df.select_dtypes(include=['float64', 'int64']).columns:
    if df[col].isnull().sum() > 0:
        df[col].fillna(df[col].mean(), inplace=True)

# Filling missing values in categorical columns with mode
for col in df.select_dtypes(include=['object']).columns:
    if df[col].isnull().sum() > 0:
        df[col].fillna(df[col].mode()[0], inplace=True)

---- provide your text answer here ----
Data imputation is the process of filling in missing or invalid values so the dataset remains complete and suitable for analysis. In this case, I followed best practices by imputing numeric variables with their mean values, which preserves the overall distribution of continuous data without biasing towards extreme values. For categorical variables, I used the mode, as it ensures missing entries are replaced with the most frequent and representative category. This strategy keeps the dataset both consistent and meaningful, allowing the models to learn from patterns without being distorted by gaps in the data.

<hr style="width:25%;margin-left:0;"> 

**Q2**. 
- Print `value_counts()` of the 'State' column and add a dummy variable named 'State_NSW' to `df` using `get_dummies()` (3 marks)
- Carefully explain what the values of the new variable 'State_NSW' mean (2 mark)
- Make sure the variable 'State' is deleted from the original dataframe   

(Total: 5 marks)  

In [7]:
# ---- provide your code here -----
# Checking the frequency distribution of the 'State' column
print(df['State'].value_counts())

# Creating dummy variables for 'State' using get_dummies
state_dummies = pd.get_dummies(df['State'], prefix='State', dtype=int)

# Keeping only the 'State_NSW' dummy
df['State_NSW'] = state_dummies['State_NSW']

# Dropping the original 'State' column
df.drop(columns=['State'], inplace=True)

State
NSW    2113
WA      887
Name: count, dtype: int64


---- provide your text answer here ----
The frequency distribution shows that most observations belong to NSW, while a smaller proportion belong to WA. The new dummy variable State_NSW takes the value 1 if the state is NSW and 0 if the state is WA. This transformation encodes the categorical state information into a numerical format, making it suitable for predictive modeling, while ensuring no observations were lost when the original State column was removed.

<hr style="width:25%;margin-left:0;"> 

**Q3**. Print `value_counts()` of the 'Whse' column and *carefully* comment on what you notice in relation to the definition of this variable. 

(Total: 5 marks) 

In [8]:
# ---- provide your code here -----
# Checking the frequency distribution of 'Whse' column
print(df['Whse'].value_counts())

Whse
1N1    1318
1W0     887
1N0     795
Name: count, dtype: int64


---- provide your text answer here ----
The Whse column represents warehouse codes that identify different storage locations. The distribution shows that most items are stored in warehouse 1N1 (1318 records), followed by 1W0 (887 records) and 1N0 (795 records). This confirms that Whse is a categorical variable describing warehouse sites rather than a numeric measure, consistent with the dataset definition.

<hr style="width:25%;margin-left:0;"> 

**Q4**. 

- Apply `get_dummies()` to 'Whse' feature and add dummy variables 'Whse_1N0', 'Whse_1N1', 'Whse_1W0' to `df`. (5 marks)   
- *Carefully consider* how to allocate all the values of 'Whse' across these 3 newly created features (5 marks)
    - Do not delete observations 
    - Do not assume that the anomaly are missing observations      
    - Explain what decision you made
- Make sure that 'Whse' is deleted from `df`   

(Total: 10 marks)   

In [9]:
# ---- provide your code here -----
# Applying get_dummies() to 'Whse' column to create dummy variables
whse_dummies = pd.get_dummies(df['Whse'], prefix='Whse', dtype=int, drop_first=True)

df = df.join(whse_dummies)

# Explicitly drop the original 'Whse' column (already handled by get_dummies, but done here for clarity)
df.drop(columns=['Whse'], inplace=True)


---- provide your text answer here ----
When I used get_dummies() to the Whse column created dummy variables for each warehouse, but with drop_first=True one category was dropped to serve as the baseline. Since the dataset only contains warehouses 1N0, 1N1, and 1W0, the dummy variables now indicate membership relative to the dropped category. For example, if Whse_1N1 = 1, the observation belongs to warehouse 1N1; if all dummy variables equal 0, the observation belongs to the baseline warehouse. The original Whse column was deleted to avoid duplication.

<hr style="width:25%;margin-left:0;"> 

**Q5**. In the column 'Item Type', convert the values of all items that are FG --- into "Finished Goods", RM --- into "Raw Materials", keep "WIP Manufactured", while combine all others like [Subcontract (9), Customer Supplied (08), and  zItems] into type 'Other' so you have Four Item types **Finished Goods**, **Raw Materials**, **WIP Manufactured**, and **Other**. 

Now Encode all the remaining non numeric features using appropriate encoding method presented in class

(Total: 5 marks)  

In [10]:
# ---- provide your code here -----
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder

df.loc[df['Item Type'].str.startswith("FG"), 'Item Type'] = "Finished Goods"
df.loc[df['Item Type'].str.startswith("RM"), 'Item Type'] = "Raw Materials"
df.loc[df['Item Type'] == "WIP Manufactured (2)", 'Item Type'] = "WIP Manufactured"
df.loc[~df['Item Type'].isin(["Finished Goods", "Raw Materials", "WIP Manufactured"]), 'Item Type'] = "Other"


item_dummies = pd.get_dummies(df['Item Type'], prefix='ItemType', dtype=int, drop_first=True)

df = df.join(item_dummies)

# Explicitly dropping the original 'Item Type' column (already handled by get_dummies, but done here for clarity)
df.drop(columns=['Item Type'], inplace=True)


baseunit_dummies = pd.get_dummies(df['Base Unit'], prefix='BaseUnit', dtype=int, drop_first=True)

df = df.join(baseunit_dummies)

# Explicitly dropping the original 'Base Unit' column (already handled by get_dummies, but done here for clarity)
df.drop(columns=['Base Unit'], inplace=True)

businessarea_dummies = pd.get_dummies(df['Business Area'], prefix='BusinessArea', dtype=int, drop_first=True)

df = df.join(businessarea_dummies)

# Explicitly dropping the original 'Business Area' column (already handled by get_dummies, but done here for clarity)
df.drop(columns=['Business Area'], inplace=True)


abc_dummies = pd.get_dummies(df['ABC Class'], prefix='ABCClass', dtype=int, drop_first=True)

df = df.join(abc_dummies)

# Explicitly drop the original 'Base Unit' column (already handled by get_dummies, but done here for clarity)
df.drop(columns=['ABC Class'], inplace=True)


# Label-encode target variable 'Dead stock'
le = LabelEncoder()
df['Dead stock'] = le.fit_transform(df['Dead stock'])


---
---

**Problem 3** Preparing X and y arrays (Total Marks: 10)

**Q1**. 

- Create a numpy array `y` from the first 80% observations of `Dead stock` column from `df` (2.5 marks)   
- Create a numpy array `X`  from the first 80% observations of all the remaining variables in `df` (2.5 marks)   

(Total: 5 Marks)

In [11]:
# ---- provide your code here -----
import numpy as np

#  define cutoff index for first 80%
split_index = int(len(df) * 0.8)

# create target array y (Dead stock)
y = df['Dead stock'].iloc[:split_index].to_numpy()

# create feature array X (all other variables except Dead stock)
X = df.drop(columns=['Dead stock']).iloc[:split_index].to_numpy()

<hr style="width:25%;margin-left:0;"> 

**Q2**. 

- Use an appropriate `sklearn` library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 80% train and 20% test datasets (2.5 marks) 
    - Set random_state to 31 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data to mean zero and variance one using an approapriate `sklearn` library (2.5 marks)   


(Total: 5 marks) 

In [12]:
# ---- provide your code here -----
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Splitting the dataset into 80% train and 20% test with stratification
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size = 0.2, random_state = 31, stratify = y
)

# Standardizing the data (mean = 0, variance = 1)
sc = StandardScaler()
sc.fit(X_train)

X_train_scaled = sc.transform(X_train)
X_test_scaled = sc.transform(X_test)

---
---

**Problem 4**. Training Models and Interpretation (Total Marks: 30)


**Q1**. 

- Train one linear classifier we studied in class using standardised data (6 marks)
- Compute and print training and test dataset accuracies (4 marks)

(Total: 10 marks)   

In [13]:
# ---- provide your code here -----
from sklearn.linear_model import LogisticRegression

# Train logistic regression classifier
lr = LogisticRegression(C=1, random_state=31, solver='lbfgs', multi_class='ovr')
lr.fit(X_train_scaled, y_train)

# Compute and print accuracies
print("Training Accuracy:", lr.score(X_train_scaled, y_train))
print("Test Accuracy:", lr.score(X_test_scaled, y_test))

Training Accuracy: 0.9963541666666667
Test Accuracy: 0.9958333333333333


/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(


<hr style="width:25%;margin-left:0;"> 

**Q2.**

- Train one nonlinear classifier we studied in class on the same dataset (6 marks)
- Compute and print training and test dataset accuracies (4 marks)


(Total: 10 marks)  

In [15]:
# ---- provide answer here -----
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import SelectFromModel

# Training a nonlinear classifier (Random Forest) on the same train set
#    RF does not require standardisation; X_train (unscaled) is fine.
rf = RandomForestClassifier(n_estimators=25, criterion='gini', random_state=31, n_jobs=-1)
rf.fit(X_train, y_train)

# Computing and printing training and test accuracies
print("Training Accuracy :", rf.score(X_train, y_train))
print("Test Accuracy :", rf.score(X_test, y_test))

Training Accuracy : 1.0
Test Accuracy : 0.9979166666666667


<hr style="width:25%;margin-left:0;"> 

**Q3**. 

- Comment on the accuracy results obtained from the two classifiers (6 marks)
- Based on our investigation into Dead stock predictions in this assignment, which model would you recommend and why? (4 marks)
- What **feature(s)** do you believe has/have higest impact on predicting the outcome variable? Why?


(Total: 10 marks)     


---- provide your text answer here ----
Comment on the Accuracy: 

Logistic Regression achieved consistently high accuracy, with training around 99.6% and test accuracy 99.5%, and showed very little variation across C values from 0.5 to 10. This stability indicates that changes in regularization strength had minimal impact, reflecting the dataset’s near-linear separability.
Random Forest also performed extremely well, reaching training accuracy of 100% and test accuracy between 99.79% and 100%. With a smaller number of trees (n_estimators=10), it already achieved 99.95% training and 99.79% test accuracy. Increasing the number of trees to 500 provided slightly more stable results but the combination of low estimators and entropy brought the best results. I believe the best result was achieved with n_estimators set to 10 and criterion set to entropy, where the training accuracy reached 99.94% and the test accuracy was 100%

Recommended Model: 
Based on our investigation, I would recommend Random Forest for predicting Dead stock. While Logistic Regression performed very well, achieving around 99.6% training accuracy and 99.6% test accuracy, Random Forest achieved even higher scores with 100% training accuracy and about 99.8% test accuracy. This consistent outperformance, even by a small margin, shows that Random Forest was better able to capture the complex interactions among features such as stock age, turnover, and ABC categories.

The best configuration was observed with n_estimators = 10 and criterion = entropy, where the model reached 99.94% training accuracy and a perfect 100% test accuracy. Increasing the number of trees (e.g., 500) improved stability further, though it introduced a mild risk of overfitting.

Therefore, Random Forest is the recommended model. It not only delivers the strongest predictive performance but also highlights the most influential features, supporting both accurate forecasting and business decision-making.

High Impact Features: 
The features with the greatest impact on predicting Dead stock are “Over 2 years Qty” , “ABC Class",  “% of over 2 year”. These variables reflect both the age of stock and its ABC classification, which are key signals of whether items are likely to become unsellable. In particular, ABC Class is critical because the categories “Not Purchasing” and “End of Life” almost always correspond to Dead stock = 1, making them the clearest indicators of risk. Similarly, “Over 2 years Qty” and “% of over 2 year” measure how long inventory has been idle, and older stock is far less likely to sell. Supporting features such as Inventory Turn and Over 3 Years Quantity also reinforce the role of turnover speed and long-term stock aging. Overall, the combination of stock age and ABC Class—especially “Not Purchasing” and “End of Life”—emerges as the most influential driver of dead stock prediction, consistent with business logic that older and low-priority items are most at risk of becoming unsellable.





---
---

## Marking Criteria

To achieve a perfect score, your solutions must adhere to the criteria outlined below:

- Ensure that all numerical answers are accurate.
- Utilize the exact Python functions and libraries specified within the assignment instructions.
- For any written responses, provide accurate information, articulated in clear, complete sentences.
- Do not add extra cells beyond what is provided in the notebook.
- Do not print output with your code unless explicitly instructed to do so.
- Maintain a clean and organised notebook layout that is easy to follow.
- Marks will be deducted for not following the above instructions.
    
---
---